### ISSUE TITLE
Finalise stocktake making invalid stock adjustments (into negative on ledger)

### ISSUE THREAD

**Issue by Chris-Petty** [sussol/mobile#629]
Build Number: 2.1.0-rc8

Description: 
Through a fun sequence of editing you can make stocktakes that show a negative ledger total for an item but show's "fine" on mobile. Both are wrong. Fortunately it's not totally easy to achieve.

Reproducible: Yes

Reproduction Steps:
1. Make a new stocktake with an item
2. Make sure it has some quantity for snapshot (say 100)
3. Make a customer invoice for the same item and issue some amount (say 100)
4. Go back to the stocktake and set the actual quantity to 0. You cannot finalise because it'd make negative stock.
5. Set actual quantity >= 100
6. Edit actual quantity to 0 AND hit finalise before anything else
7. Confirm and viola, you have messed the ledger.

In mSupply mobile, it'll say there is 0 stock. In mSupply Desktop for that store it'll say 0 stock too, but by the ledger it should have -100 stock.

(I've done it a couple times here, so -250)
![image](https://user-images.githubusercontent.com/7684221/37379532-bb65fdf0-2798-11e8-88ac-3ae3b1cfeac4.png)

Comments: 🙈

**Comment by anildahalsussol** [comment]
Yah I also faced this problem in medical action myanmar with negative stock which is creating ledger problem. Sorry i didnot made the issue but i informed to andrei about it.

**Comment by andreievg** [comment]
@Chris-Petty 
I think the really issue is in react-native-data table and what happens when cell loses focus.  We can also finalise CustomInvoice/Stocktake with 0 quantities, i.e. set quantity to something other then zero, then back to zero then immediately finalise... 

I thought this PR fixed it https://github.com/sussol/react-native-data-table/pull/9/files
I've tried to explain here: https://github.com/sussol/mobile/issues/427
Obviously still an issue. 

I like your fix here: https://github.com/sussol/mobile/pull/634 , I think it can be extended, after checking 'checkForFinaliseError' and before record.finalise.. call checkForFinalisationError again.

**Comment by andreievg** [comment]
I had a further look into this issue, first assumption was incorrect, it's nothing to do with realm.. It seems that onEndEditing is called after the button click, it's called just after the modal is shown. If you create a simple react native app with just a button and a text input, when button is pressed the text is still in focus.  We can call Keyboard.dismiss() to un-focus the text input, and only after that try to open the finalisation modal. I've tested the code below, it works, but the timeout of 1 second might not be sufficient in all cases: 
```
// mSupplyMobileApp
  renderFinaliseButton = () => (
    <FinaliseButton
      isFinalised={this.props.finaliseItem.record.isFinalised}
      onPress={() => {
        Keyboard.dismiss();
        setTimeout(() => this.setState({ confirmFinalise: true }), 1000);
      }}
    />
  )
```

**Comment by Chris-Petty** [comment]
Yea I also tested keyboard.dismiss to try and get onEndEditing to run before the modal did it's thing, but it wasn't quick enough. The setTimeout is a bit of a hack/dirty I thought and potentially flawed as you mentioned.

I did try to think if there was a way to use `database.isInTransaction` and async/promise to do it a bit smarter but was in a bit of a hurry so made a PR of what I had done.

**Comment by Chris-Petty** [comment]
Fixed by #634 

Further discussions at #638
